File: Data Structures & Algorithms/design-word-search-data-structure/submission-0.cpp
class Trie {

public:
    Trie *node[26];
    bool isEnd;

    Trie() {
        isEnd = false;
        for (int i = 0; i < 26; i++) {
            node[i] = NULL;
        }
    }
};

class WordDictionary {
public:
    Trie *trie;
    WordDictionary() {
        trie = new Trie();
    }
    
    void addWord(string word) {
        Trie *ptr = trie;
        for (int i = 0; i < word.length(); i++) {
            if (!ptr->node[word[i] - 'a']) {
                ptr->node[word[i] - 'a'] = new Trie();
                ptr = ptr->node[word[i] - 'a'];
            } else {
                ptr = ptr->node[word[i] - 'a'];
            }
        }
        ptr->isEnd = true;
    }
    
    bool search(string word) {
        return find(trie, word, 0);
    }

    bool find(Trie *node, string &word, int index) {
        if (index == word.length()) {
            if (node->isEnd) {
                return true;
            } else {
                return false;
            }
        }
        if (word[index] == '.') {
            for (int i = 0; i < 26; i++) {
                if (node->node[i]) {
                    if (find(node->node[i], word, index+1)) {
                        return true;
                    }
                }
            }
        } else {
            if (node->node[word[index] - 'a']) {
                return find(node->node[word[index] - 'a'], word, index+1);
            }
        }
        return false;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: Data Structures & Algorithms/design-twitter-feed/submission-0.java
class Tweet implements Comparable<Tweet> {
    private final Integer tweetId;
    private final Integer userId;
    private final Integer timestamp;

    public Tweet(Integer tweetId, Integer userId, Integer timestamp) {
        this.tweetId = tweetId;
        this.userId = userId;
        this.timestamp = timestamp;
    }

    public Integer getTimestamp() {
        return timestamp;
    }

    public Integer getTweetId() {
        return tweetId;
    }

    public Integer getUserId() {
        return userId;
    }

    @Override
    public int compareTo(final Tweet other) {
        return Integer.compare(other.getTimestamp(), this.getTimestamp());
    }
}

public class Node {
    Tweet tweet;
    int userId;
    int index;

    Node(Tweet tweet, int userId, int index) {
        this.tweet = tweet;
        this.userId = userId;
        this.index = index;
    }
}

class Twitter {

    private final Map<Integer, List<Tweet>> userTweets;
    private final Map<Integer, List<Tweet>> userFeed;
    private final Map<Integer, Set<Integer>> userFollowings;
    private final Map<Integer, Set<Integer>> userFollowers;
    private final AtomicInteger tweetNumber;

    public Twitter() {
        userTweets = new HashMap<>();
        userFeed = new HashMap<>();
        userFollowings = new HashMap<>();
        userFollowers = new HashMap<>();
        tweetNumber = new AtomicInteger(0);
    }
    
    public void postTweet(int userId, int tweetId) {
        final Tweet tweet = new Tweet(tweetId, userId, tweetNumber.getAndIncrement());
        userTweets.computeIfAbsent(userId, k -> new ArrayList<>()).add(tweet);
    }
    
    public List<Integer> getNewsFeed(int userId) {
        PriorityQueue<Node> feed = new PriorityQueue<>((a, b) -> Integer.compare(b.tweet.getTimestamp(), a.tweet.getTimestamp()));
        Set<Integer> followings = userFollowings.getOrDefault(userId, new HashSet<>());
        for (Integer followingUserId: followings) {
            List<Tweet> followingUserTweets = userTweets.getOrDefault(followingUserId, new ArrayList<>());
            if (!followingUserTweets.isEmpty()) {
                feed.add(new Node(followingUserTweets.getLast(), followingUserId, followingUserTweets.size() - 1));
            }
        }
        if (!userTweets.getOrDefault(userId, new ArrayList<>()).isEmpty()) {
            feed.add(new Node(userTweets.get(userId).getLast(), userId, userTweets.get(userId).size() - 1));
        }

        int count = 0;
        List<Integer> response = new ArrayList<>();

        while (count < 10 && !feed.isEmpty()) {
            Node current = feed.poll();
            response.add(current.tweet.getTweetId());

            if (current.index > 0) {
                Tweet lastTweet = userTweets.get(current.userId).get(current.index - 1);
                feed.add(new Node(lastTweet, current.userId, current.index - 1));
            }

            count++;
        }

        return response;
     }
    
    public void follow(int followerId, int followeeId) {
        if (followerId == followeeId) {
            return;
        }
        userFollowings.computeIfAbsent(followerId, k -> new HashSet<>()).add(followeeId);
        userFollowers.computeIfAbsent(followeeId, k -> new HashSet<>()).add(followerId);
    }
    
    public void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) {
            return;
        }
        userFollowings.computeIfAbsent(followerId, k -> new HashSet<>()).remove(followeeId);
        userFollowers.computeIfAbsent(followeeId, k -> new HashSet<>()).remove(followerId);
        final List<Tweet> followerNewsFeed = userFeed.getOrDefault(followerId, new ArrayList<>());
        followerNewsFeed.removeIf(tweet -> tweet.getUserId() == followeeId);
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */